service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,372 @@
+     * got result by {@link Criteria} will be sorted according to {@link #ORDER_KEY} parameter. If {@link #ORDER_KEY} parameter is not
+     * present or not equal "DESC" then sorting by ascending order.
+     *
+     * @param reqParams parameters of a request.
+     * @return List of GiftCertificateDao.
+     * @since 1.0
+     */
+    @Override
+    public List<GiftCertificateDto> findAllForQuery(Map<String, String[]> reqParams) {
+
+        CriteriaFactoryResult<GiftCertificate> criteriaWithParams = criteriaFactory.getCriteriaWithParams(reqParams);
+
+        List<GiftCertificate> foundCertificates = giftCertificateDao.getBy(criteriaWithParams);
+
+        if (foundCertificates.isEmpty()) {
+            throw new GiftCertificateNotFoundException("Can't find gift certificates");
+        }
+
+        if (reqParams.containsKey(SORT_FIELDS_KEY)) {
+            String[] sortFields = reqParams.get(SORT_FIELDS_KEY);
+            String[] orders = reqParams.get(ORDER_KEY);
+
+            foundCertificates = sortingHelper.getSorted(sortFields, orders[0], foundCertificates);
+        }
+
+        return foundCertificates.stream()
+                .map(certificate -> modelMapper.map(certificate, GiftCertificateDto.class))
+                .peek(this::fillCertificateDtoWithTags)
+                .collect(Collectors.toList());
+    }
+
+
+    /**
+     * This method gets GiftCertificate entity from dao layer with given id and converts it to GiftCertificateDto.
+     * The method also calls {@link #fillCertificateDtoWithTags(GiftCertificateDto)} for filling gotten GiftCertificateDto
+     * with list of TagDto.
+     *
+     * @param id id of necessary entity.
+     * @return GiftCertificateDto with id and tags.
+     * @throws GiftCertificateNotFoundException if there is no entity with given id in database.
+     * @since 1.0
+     */
+    @Override
+    public GiftCertificateDto findById(long id) {
+        Optional<GiftCertificate> foundCertificateOpt = giftCertificateDao.getById(id);
+
+        GiftCertificate foundCertificate = foundCertificateOpt.orElseThrow(() ->
+                new GiftCertificateNotFoundException(String.format("Can't find a certificate with id: %d", id)));
+
+        GiftCertificateDto giftCertificateDto = modelMapper.map(foundCertificate, GiftCertificateDto.class);
+        fillCertificateDtoWithTags(giftCertificateDto);
+        return giftCertificateDto;
+    }
+
+
+    /**
+     * This method set date fields for GiftCertificate entity and saves it in database. Also the method link passed tags with the GiftCertificate
+     * entity and, if the tag doesn't exist in db, creates it.
+     *
+     * @param certificate GiftCertificate entity without date fields and id for saving.
+     * @return GiftCertificateDto with id and tags.
+     * @since 1.0
+     */
+    @Override
+    @Transactional
+    public GiftCertificateDto save(GiftCertificateDto certificate) {
+
+        String certificateName = certificate.getName();
```

Comment:
This logic may be moved to CertificatesValidator

Reply:
I moved the code of checking an existence or non-existence of an entity to EntityValidator<T> class.
commit 0c4dd35.